src/gridrsp.jl

```diff
@@ -305,6 +370,18 @@ function least_cost_kl_divergence(grsp::GridRSP, target::Tuple{Int,Int})
     return reshape(div, grsp.g.nrows, grsp.g.ncols)
 end
 
+
+@kwdef struct ConnectedHabitat{CV,DT,DV} <: RSPOperation
```

Comment:
Yes GridRSP id pretty much hidden now in this PR. You call `compute` on a `Problem` and `Grid` and GridRSP is internal.

I'm working towards preallocating the whole problem up front like LinearSolve does  but most of the code is unfortunately not written in that style. The first step will be to separate out non-allocating `!` methods for everything so the workspace arrays can be passed in.

Reply:
Also had a good discussion with @ninbrm about refactoring distances and fixing the abstract type structure to better represent the range of problems. Theta really doesn't need to be a parameter but the fact that distance functions are curreently used like functions rather than constructed objects currently makes it awkward to put parameters in them. So we should change them to be types like you describe.

I'll implement this new abstract type hierarchy in the next iteration of this PR and we can discuss from there

Reply:
Great, looking forward to that!

Reply:
Ok the new abstractions are kind of implemented now :)

It works in the landmarks.jmd example at least, but needs some more polish and though. Eventually we can refactor some of the complexity away but for now Im trying to build it on top of the current methods without changing them too much, so the old tests still run and we can use them to check the new methods.